# 1. Gradient Descent

## Simple Definition

**Gradient Descent is an optimization method used to reduce the model's loss by changing the weights step by step.**

In simple words:

> **Gradient Descent helps the model find better weights so that its predictions become more accurate.**

---

## Easy Explanation

Imagine you are standing on a **hill** and want to reach the lowest point.

You don't know exactly where the lowest point is.

So you:

1. Look at the slope around you.
2. Find which direction goes downward.
3. Take a small step in that direction.
4. Check again.
5. Repeat until you reach a low point.

In machine learning:

```text
Hill              → Loss
Your position     → Model weights
Downhill direction → Gradient
Step size         → Learning Rate
Lowest point      → Minimum Loss
```

So:

```text
Gradient Descent
       ↓
Find direction that reduces loss
       ↓
Change weights
       ↓
Loss becomes smaller
       ↓
Repeat
```

---

## Why Do We Need Gradient Descent?

A neural network has weights.

For example:

```text
Weight 1 = 0.5
Weight 2 = 0.8
Weight 3 = -0.2
```

Initially, these weights may not be good enough.

The model makes a prediction:

```text
Actual Price      = ₹50 lakh
Predicted Price   = ₹65 lakh
```

There is an error.

The **loss function** tells us how large that error is.

Gradient Descent helps answer:

> **"How should I change the weights to reduce this error?"**

---

## Mathematical Intuition

The basic weight update is:

```text
New Weight = Old Weight - Learning Rate × Gradient
```

For example:

```text
Old Weight    = 5
Gradient      = 2
Learning Rate = 0.1
```

Then:

```text
New Weight
= 5 - (0.1 × 2)
= 5 - 0.2
= 4.8
```

So:

```text
Old Weight → 5.0
New Weight → 4.8
```

The weight moved in the direction that reduces the loss.

---

## How Gradient Descent Works

The basic process is:

```text
1. Start with weights
        ↓
2. Make prediction
        ↓
3. Calculate loss
        ↓
4. Calculate gradient
        ↓
5. Update weights
        ↓
6. Repeat
```

For a neural network:

```text
Input
  ↓
Weights
  ↓
Prediction
  ↓
Loss
  ↓
Gradient
  ↓
Gradient Descent
  ↓
Updated Weights

---

## Real-World Example

Imagine an **AI model predicting house prices**.

The model predicts:

```text
Actual house price     = ₹50 lakh
Predicted house price  = ₹70 lakh
```

The model has a large error.

Gradient Descent looks at the error and calculates how the model's weights should change.

For example:

```text
House Size weight
Bedrooms weight
Location weight
Age weight
```

The weights are adjusted:

```text
Old weights
    ↓
Calculate gradient
    ↓
Update weights
    ↓
New weights
    ↓
Better prediction
```

After many training steps, the model may produce:

```text
Actual price     = ₹50 lakh
Predicted price  = ₹52 lakh
```

The loss has decreased.

---

## Advantages

* Simple optimization concept
* Helps reduce training loss
* Used as the foundation for many neural-network optimizers
* Works with large numbers of model parameters

---

## Limitations

* Learning rate needs to be chosen carefully.
* Very small learning rate → training can be slow.
* Very large learning rate → training can become unstable.
* Basic Gradient Descent may take many steps to reach a good solution.

---

## When to Use

Gradient Descent is the **basic idea behind training many machine-learning and deep-learning models**.

You don't always implement basic Gradient Descent manually. Optimizers such as:

```text
SGD
SGD + Momentum
RMSprop
Adam
```

build on or improve the basic weight-update idea.

---

## Key Points

```text
Gradient Descent
        ↓
Reduces Loss
        ↓
By changing Weights
        ↓
Step by Step
```

Remember these 3 things:

**Gradient** → tells the direction.

**Learning Rate** → tells how big the step is.

**Gradient Descent** → uses them to update the weights.


In [1]:

## Python Example

#Let's use a very simple example.

#Our goal is to find the value of `weight` that makes:

#weight → 0


#The loss is:

#```text
#Loss = weight²
#```

#So if:
#
#```text
#weight = 5
#```

#loss is:

#```text
#25
#```

#If:

#```text
##weight = 1
##```

#loss is:


#Therefore, moving the weight toward `0` reduces the loss.


weight = 5.0
learning_rate = 0.1

for step in range(5):

    # Calculate gradient
    gradient = 2 * weight

    # Update weight
    weight = weight - learning_rate * gradient

    print("Step:", step + 1)
    print("Weight:", round(weight, 4))
    print()


Step: 1
Weight: 4.0

Step: 2
Weight: 3.2

Step: 3
Weight: 2.56

Step: 4
Weight: 2.048

Step: 5
Weight: 1.6384



# 2. Batch Gradient Descent

## Simple Definition

**Batch Gradient Descent calculates the gradient using the entire training dataset before updating the model's weights.**

In simple words:

> **It looks at all training data, calculates the overall direction, and then updates the weights once.**

---

## Easy Explanation

Suppose we have **5 training records**:

```text
Record 1
Record 2
Record 3
Record 4
Record 5
```

Batch Gradient Descent uses **all 5 records together**:

```text
All 5 records
     ↓
Make predictions
     ↓
Calculate loss
     ↓
Calculate gradient
     ↓
Update weights
```

Then it starts the next iteration with the updated weights.

---

## How Is It Different From Basic Gradient Descent?

Gradient Descent is the general idea:

> Move the weights in a direction that reduces loss.

**Batch Gradient Descent** specifies **how much data is used for each update**:

```text
Batch Gradient Descent
→ Entire training dataset
→ One weight update
```

For example, if you have:

```text
10,000 training records
```

Batch Gradient Descent uses:

```text
10,000 records → calculate gradient → update weights
```

---

## Mathematical Intuition

Suppose we have three training examples.

Each example produces a gradient:

```text
Record 1 → Gradient = 2
Record 2 → Gradient = 4
Record 3 → Gradient = 3
```

Batch Gradient Descent calculates the average:

```text
Average Gradient
= (2 + 4 + 3) / 3
= 3
```

Then it updates the weight using that overall gradient.

```text
New Weight
= Old Weight - Learning Rate × Average Gradient
```

The important idea is:

> **All training examples contribute to one update.**

---

## Step-by-Step Example

Imagine a house-price model with 4 houses:

```text
House 1 → ₹40 lakh
House 2 → ₹50 lakh
House 3 → ₹60 lakh
House 4 → ₹70 lakh
```

The model makes predictions for **all 4 houses**.

```text
All houses
    ↓
Calculate prediction errors
    ↓
Calculate overall loss
    ↓
Calculate gradient
    ↓
Update weights
```

Only after processing all four houses does the model update its weights.

---
`

---

## Real-World Example

Imagine an **online shopping company** wants to predict whether customers will purchase a product.

The company has:

```text
100,000 customer records
```

With Batch Gradient Descent:

```text
100,000 customers
       ↓
Calculate predictions
       ↓
Calculate overall loss
       ↓
Calculate gradient
       ↓
Update model weights
```

Then the next training iteration processes all 100,000 records again.

---

## Advantages

### 1. Stable Updates

Because the gradient is calculated using all training data, updates are generally less noisy.

### 2. Predictable Training

The model follows a more consistent direction during optimization.

### 3. Simple to Understand

The complete dataset is used for each update.

---

## Limitations

### 1. Slow for Large Datasets

If you have millions of records, processing everything before every update can be slow.

### 2. Higher Memory Requirement

The complete dataset needs to be handled during the calculation.

### 3. Fewer Updates

Compared with mini-batch or SGD, you get fewer weight updates during an epoch.

---

## When to Use

Batch Gradient Descent can be useful when:

* Dataset is relatively small
* You want stable gradient calculations
* Memory is sufficient
* Training speed is not a major concern

For very large neural-network datasets, **mini-batch training is usually more practical**.

---

## Easy Comparison

Suppose you have **1,000 training records**.

```text
Batch Gradient Descent
1000 records → Update
1000 records → Update
1000 records → Update
```

Compare:

```text
SGD
1 record → Update
1 record → Update
1 record → Update
```

And:

```text
Mini-Batch
32 records → Update
32 records → Update
32 records → Update
```

So the main difference is **how many training examples are used for one weight update**.

---

## Key Points

Remember:

```text
Batch Gradient Descent
        ↓
Uses ALL training data
        ↓
Calculate gradient
        ↓
Update weights
        ↓
Repeat
```

### One-line memory trick:

> **Batch = Entire dataset for one update.**


In [2]:

## Python Example

import numpy as np

# House sizes
X = np.array([1000, 1500, 2000, 2500])

# House prices
y = np.array([200, 300, 400, 500])

# Initial weight
weight = 0.0

learning_rate = 0.0000001

for epoch in range(5):

    # Make predictions using all training records
    predictions = weight * X

    # Calculate errors
    errors = predictions - y

    # Calculate gradient using ALL records
    gradient = np.mean(2 * X * errors)

    # Update weight
    weight = weight - learning_rate * gradient

    print(
        "Epoch:", epoch + 1,
        "Weight:", round(weight, 6)
    )



Epoch: 1 Weight: 0.135
Epoch: 2 Weight: 0.178875
Epoch: 3 Weight: 0.193134
Epoch: 4 Weight: 0.197769
Epoch: 5 Weight: 0.199275


In [ ]:

---

## Output

You may see the weight gradually changing:

```text
Epoch: 1 Weight: 0.23
Epoch: 2 Weight: 0.35
Epoch: 3 Weight: 0.42
Epoch: 4 Weight: 0.46
Epoch: 5 Weight: 0.48
```

The exact values depend on the learning rate.

The important part is that **all four houses are used before each weight update**.

---

## What This Code Is Doing

Look at this line:

```python
predictions = weight * X
```

`X` contains all four house sizes.

So the model predicts for:

```text
1000
1500
2000
2500
```

Then:

```python
gradient = np.mean(2 * X * errors)
```

calculates the average gradient from **all training records**.

Then:

```python
weight = weight - learning_rate * gradient
```

updates the weight.

So:

```text
All Data
   ↓
One Gradient
   ↓
One Weight Update
``

# 3. Stochastic Gradient Descent (SGD)

## Simple Definition

**Stochastic Gradient Descent (SGD) updates the model's weights using one training example at a time.**

In simple words:

> **Instead of waiting for the whole dataset, the model learns from one record and immediately updates its weights.**

---

## Easy Explanation

Suppose we have 5 training records:

```text
Record 1
Record 2
Record 3
Record 4
Record 5
```

SGD works like this:

```text
Record 1
   ↓
Calculate gradient
   ↓
Update weights

Record 2
   ↓
Calculate gradient
   ↓
Update weights

Record 3
   ↓
Calculate gradient
   ↓
Update weights
```

So the model updates its weights **after every individual training example**.

---

## Batch vs SGD

This is the most important difference.

### Batch Gradient Descent

```text
1000 records
     ↓
Calculate gradient
     ↓
One update
```

### SGD

```text
Record 1 → Update
Record 2 → Update
Record 3 → Update
...
Record 1000 → Update
```

So:

> **Batch = all data → one update**

> **SGD = one record → one update**

---

## Why Is SGD Useful?

Imagine you have:

```text
10 million training records
```

With Batch Gradient Descent:

```text
10 million records
       ↓
Calculate gradient
       ↓
Update weights
```

You have to process the whole dataset before getting an update.

SGD can instead do:

```text
Record 1 → Update
Record 2 → Update
Record 3 → Update
...
```

The model starts learning immediately.

---

## Mathematical Intuition

The basic update is still:

```text
New Weight = Old Weight - Learning Rate × Gradient
```

The difference is **where the gradient comes from**.

With SGD:

```text
Gradient = gradient from one training example
```

For example:

```text
Old Weight = 5
Gradient = 2
Learning Rate = 0.1
```

Then:

```text
New Weight
= 5 - (0.1 × 2)
= 4.8
```

Then the next training example calculates a new gradient and updates the weight again.

---

## How SGD Works

The complete process:

```text
Training Record 1
       ↓
Prediction
       ↓
Loss
       ↓
Gradient
       ↓
Update Weights
       ↓
Training Record 2
       ↓
Prediction
       ↓
Loss
       ↓
Gradient
       ↓
Update Weights
```

This continues through the training dataset.



## Real-World Example

Imagine an **online shopping recommendation system**.

Every time a customer interacts with the website:

```text
Customer 1
   ↓
Clicks product
   ↓
Model learns
   ↓
Weights updated

Customer 2
   ↓
Buys product
   ↓
Model learns
   ↓
Weights updated
```

The model doesn't necessarily need to wait until millions of customers have interacted before making an update.

This is the basic idea behind learning from individual examples.

---

## Advantages

### 1. Fast Updates

The model updates after every training example.

### 2. Useful for Large Datasets

You don't need to calculate one gradient using the entire dataset.

### 3. Lower Memory Requirement

You can process data one example at a time.

### 4. Can Escape Some Difficult Regions

The noisy updates can sometimes help the optimization process move away from certain local regions.

---

## Limitations

### 1. Noisy Training

The loss may go:

```text
10 → 8 → 9 → 6 → 7 → 4
```

instead of smoothly decreasing.

### 2. Can Be Unstable

A learning rate that is too large can cause large jumps.

### 3. More Frequent Updates

Because weights change after every example, the training behavior can be less predictable.

---

## When to Use

SGD can be useful when:

* Dataset is very large
* You want frequent updates
* You need memory-efficient training
* You are working with online or streaming-style learning

However, in modern neural-network training, **mini-batch gradient descent is much more commonly used than pure one-example-at-a-time SGD**.

---

## Important PyTorch Note

When you write:

```python
torch.optim.SGD(model.parameters(), lr=0.01)
```

the optimizer itself is called **SGD**.

But if your DataLoader uses:

```text
batch_size = 32
```

then the optimizer is actually receiving **mini-batches of 32 examples**, not one example at a time.

So don't confuse:

```text
SGD optimizer
```

with:

```text
Pure stochastic training using exactly 1 record
```

The PyTorch `SGD` optimizer can be used with different batch sizes.

---

## Key Points

```text
SGD
 ↓
Take one training example
 ↓
Calculate prediction
 ↓
Calculate loss
 ↓
Calculate gradient
 ↓
Immediately update weights
 ↓
Take next example
```

### Easy memory trick:

> **SGD = Learn → Update → Learn → Update**

And remember the main difference:

```text
Batch GD      → All records → Update
SGD           → 1 record → Update
Mini-Batch GD → Small group → Update
```




---

## What Is Happening Here?

This line:

```python
house_size = X[i]
```

selects **one house**.

For example:

```text
First iteration
House size = 1000
```

The model calculates:

```text
Prediction
   ↓
Error
   ↓
Gradient
   ↓
Weight update
```

Then:

```text
Second iteration
House size = 1500
```

The model again calculates a gradient and updates the weight.

So:

```text
House 1 → Update
House 2 → Update
House 3 → Update
House 4 → Update
```

---

## Output

You may see something like:

```text
Epoch: 1 Weight: 0.399
Epoch: 2 Weight: 0.482
Epoch: 3 Weight: 0.498
```

The exact values depend on the learning rate and data.

The important thing is that the weight is updated **after each individual record**.

---

## Why Is SGD Noisy?

Suppose we have:

```text
Record 1 → Gradient = +5
Record 2 → Gradient = -2
Record 3 → Gradient = +4
Record 4 → Gradient = -1
```

The direction can change from one record to another.

Therefore, the training path can look like:

```text
      ↗
   ↘
     ↗
  ↘
    ↗
```

Instead of moving smoothly toward the minimum.

This is called **noisy or fluctuating training**.

But that noise can sometimes help the optimizer move out of certain difficult regions.

---

In [4]:

## Python Example

import numpy as np

# House sizes
X = np.array([1000, 1500, 2000, 2500])

# House prices in lakhs
y = np.array([200, 300, 400, 500])

weight = 0.0

learning_rate = 0.0000001

for epoch in range(3):

    for i in range(len(X)):

        # Take ONE training example
        house_size = X[i]
        actual_price = y[i]

        # Prediction
        prediction = weight * house_size

        # Error
        error = prediction - actual_price

        # Gradient from ONE example
        gradient = 2 * house_size * error

        # Update weight immediately
        weight = weight - learning_rate * gradient

    print(
        "Epoch:", epoch + 1,
        "Weight:", round(weight, 6)
    )


Epoch: 1 Weight: 0.2044
Epoch: 2 Weight: 0.199903
Epoch: 3 Weight: 0.200002


# 4. Mini-Batch Gradient Descent

## Simple Definition

**Mini-Batch Gradient Descent updates the model's weights using a small group of training examples at a time.**

In simple words:

> **Instead of using the entire dataset or just one record, we use a small batch of records for each update.**

---

## Easy Explanation

Suppose we have **1,000 training records**.

We choose:

```text
Batch Size = 32
```

The model works like this:

```text
Records 1–32
     ↓
Calculate gradient
     ↓
Update weights

Records 33–64
     ↓
Calculate gradient
     ↓
Update weights

Records 65–96
     ↓
Calculate gradient
     ↓
Update weights
```

So the model doesn't wait for all 1,000 records.

---

## Why Do We Need Mini-Batches?

We have three approaches:

### Batch Gradient Descent

```text
1000 records
     ↓
One update
```

### SGD

```text
1 record
   ↓
One update
```

### Mini-Batch Gradient Descent

```text
32 records
   ↓
One update
```

So Mini-Batch is a **middle approach**.

```text
Batch GD       → Entire dataset
Mini-Batch GD  → Small group
SGD            → One record
```

---

## Mathematical Intuition

Suppose our batch contains 4 records.

Each record produces a gradient:

```text
Record 1 → Gradient = 2
Record 2 → Gradient = 4
Record 3 → Gradient = 3
Record 4 → Gradient = 5
```

Mini-Batch Gradient Descent can calculate the average:

```text
Average Gradient
= (2 + 4 + 3 + 5) / 4
= 3.5
```

Then the weight is updated using this batch gradient.

```text
New Weight
= Old Weight - Learning Rate × Batch Gradient
```

The important idea:

> **Multiple records contribute to one weight update.**

---

## How It Works

Suppose:

```text
Dataset = 10,000 records
Batch Size = 100
```

The training process looks like:

```text
100 records
    ↓
Gradient
    ↓
Update

Next 100 records
    ↓
Gradient
    ↓
Update

Next 100 records
    ↓
Gradient
    ↓
Update
```

This continues until all 10,000 records are processed.

---

## What Is Batch Size?

**Batch size = number of training examples used for one weight update.**

Examples:

```text
batch_size = 16
→ 16 records per update

batch_size = 32
→ 32 records per update

batch_size = 64
→ 64 records per update

batch_size = 128
→ 128 records per update
```

---

---

## Real-World Example

Imagine an **e-commerce company** has:

```text
1,000,000 customer records
```

Training on all 1 million records before every update can be expensive.

Instead, the company can use:

```text
Batch size = 64
```

Then:

```text
64 customers
     ↓
Calculate gradient
     ↓
Update model

Next 64 customers
     ↓
Calculate gradient
     ↓
Update model
```

The model gets frequent updates while still using multiple examples to calculate each update.

---

## Why Is Mini-Batch Common in Deep Learning?

Neural networks can have:

```text
Thousands
or
Millions
or
Billions
```

of parameters.

Training with the entire dataset at once can be expensive.

Using small batches allows hardware such as GPUs to process many examples efficiently.

That's why you commonly see:

```python
DataLoader(..., batch_size=32)
```

or:

```python
DataLoader(..., batch_size=64)
```

during neural-network training.

---

## Advantages

### 1. Faster Training

You don't need to wait for the entire dataset before every update.

### 2. More Stable Than Pure SGD

Using several records gives a more reliable gradient than using only one record.

### 3. Efficient for Neural Networks

Mini-batches work well with GPUs and other parallel hardware.

### 4. Good Balance

It provides a balance between:

```text
Batch GD → Stable but potentially slow
SGD → Fast updates but noisy
```

---

## Limitations

### 1. Batch Size Needs to Be Chosen

Common values include:

```text
16
32
64
128
```

But the best value depends on the problem and available hardware.

### 2. Still Has Some Noise

Unlike full-batch training, the gradient is based on only part of the dataset.

### 3. Large Batch = More Memory

A larger batch requires more memory.

---

## When to Use

Mini-Batch Gradient Descent is commonly used when:

* Training neural networks
* Working with large datasets
* Using GPUs
* You want a balance between speed and stable learning

For deep learning, **mini-batch training is the standard practical approach**.

---

## Easy Comparison

Suppose you have **1,000 records**.

```text
Batch GD
1000 → Update
1000 → Update
1000 → Update
```

```text
SGD
1 → Update
1 → Update
1 → Update
1 → Update
...
```

```text
Mini-Batch GD
32 → Update
32 → Update
32 → Update
...
```

---

## Key Points

Remember:

```text
Batch GD
→ Entire dataset → Update

SGD
→ One record → Update

Mini-Batch GD
→ Small group → Update
```

### One-line memory trick:

> **Mini-Batch = Take a small group → calculate gradient → update weights.**


In [5]:

## Python Example

import numpy as np

# House sizes
X = np.array([1000, 1500, 2000, 2500, 3000, 3500])

# House prices in lakhs
y = np.array([200, 300, 400, 500, 600, 700])

weight = 0.0

learning_rate = 0.0000001

batch_size = 2

for epoch in range(3):

    for start in range(0, len(X), batch_size):

        # Select a small batch
        X_batch = X[start:start + batch_size]
        y_batch = y[start:start + batch_size]

        # Predictions
        predictions = weight * X_batch

        # Error
        errors = predictions - y_batch

        # Calculate gradient using the batch
        gradient = np.mean(2 * X_batch * errors)

        # Update weight
        weight = weight - learning_rate * gradient

    print(
        "Epoch:", epoch + 1,
        "Weight:", round(weight, 6)
    )


Epoch: 1 Weight: 0.196203
Epoch: 2 Weight: 0.199928
Epoch: 3 Weight: 0.199999



```

---

## What Is Happening?

We have:

```text
6 training records
```

and:

```text
batch_size = 2
```

Therefore:

```text
Batch 1 → Records 1, 2
Batch 2 → Records 3, 4
Batch 3 → Records 5, 6
```

The updates happen like this:

```text
Records 1–2
    ↓
Update weight

Records 3–4
    ↓
Update weight

Records 5–6
    ↓
Update weight
```

---

## Output

You may see the weight gradually move toward the useful value:

```text
Epoch: 1 Weight: 0.45
Epoch: 2 Weight: 0.49
Epoch: 3 Weight: 0.50
```

The exact output depends on the learning rate and data.

The important thing is:

> **The weight is updated after every small batch, not after the entire dataset.**

# 5. Learning Rate

## Simple Definition

**Learning Rate controls how big a step the model takes when updating its weights.**

In simple words:

> **Learning rate tells the model how quickly it should learn.**

---

## Easy Explanation

Imagine you are walking down a hill and trying to reach the lowest point.

### Small steps

```text
↓
↓
↓
↓
```

You move safely, but slowly.

### Large steps

```text
↘
  ↘
    ↘
```

You move faster, but you might jump past the lowest point.

The learning rate works in a similar way.

```text
Small Learning Rate
→ Small weight updates

Large Learning Rate
→ Large weight updates
```

---

## Where Is Learning Rate Used?

We saw this equation in Gradient Descent:

```text
New Weight = Old Weight - Learning Rate × Gradient
```

For example:

```text
Old Weight    = 5
Gradient      = 2
Learning Rate = 0.1
```

Then:

```text
New Weight
= 5 - (0.1 × 2)
= 4.8
```

So the learning rate controls **how much the gradient affects the weight update**.

---

## What Happens With a Very Small Learning Rate?

Suppose:

```text
Learning Rate = 0.0001
```

The model makes very small changes.

```text
Weight
5.000
 ↓
4.999
 ↓
4.998
 ↓
4.997
```

Training may become very slow.

### Simple idea

> **Too small → model learns very slowly.**

---

## What Happens With a Suitable Learning Rate?

Suppose the learning rate is appropriate.

The model can move toward a lower loss efficiently:

```text
Loss
 ↓
  ↓
   ↓
    ↓
     ↓
```

Training progresses steadily.

### Simple idea

> **Suitable learning rate → efficient learning.**

---

## What Happens With a Very Large Learning Rate?

Suppose:

```text
Learning Rate = 1.0
```

The model may make very large changes.

It can jump from one side to another:

```text
Loss
 ↓     ↑
   ↓ ↑
     ↑
   ↓
```

The loss may fluctuate or even increase instead of decreasing.

### Simple idea

> **Too large → model may overshoot and become unstable.**

---

## Three Important Cases

```text
Too Small
    ↓
Very Slow Training

Good
    ↓
Steady Training

Too Large
    ↓
Unstable Training
```

---

## Mathematical Intuition

Suppose:

```text
Weight = 10
Gradient = 4
```

Let's try different learning rates.

### Learning Rate = 0.01

```text
New Weight
= 10 - (0.01 × 4)
= 9.96
```

Small change.

### Learning Rate = 0.1

```text
New Weight
= 10 - (0.1 × 4)
= 9.6
```

Bigger change.

### Learning Rate = 0.5

```text
New Weight
= 10 - (0.5 × 4)
= 8
```

Much bigger change.

So:

```text
Learning Rate
     ↓
Controls
     ↓
Size of Weight Update
```

---


## Real-World Example

Imagine you are training an AI model to predict **house prices**.

The model predicts:

```text
Actual Price     = ₹50 lakh
Predicted Price  = ₹70 lakh
```

The model needs to adjust its weights.

### Very small learning rate

The model changes the weights only slightly.

```text
₹70 lakh
   ↓
₹69.8 lakh
   ↓
₹69.6 lakh
```

Learning may be slow.

### Very large learning rate

The model may change its weights too aggressively.

```text
₹70 lakh
   ↓
₹40 lakh
   ↓
₹80 lakh
   ↓
₹30 lakh
```

Training can become unstable.

### Suitable learning rate

The model gradually moves toward better predictions.

```text
₹70 lakh
   ↓
₹62 lakh
   ↓
₹56 lakh
   ↓
₹52 lakh
   ↓
₹50 lakh
```

This is the behavior we generally want.

---

## Advantages of Choosing a Good Learning Rate

* Training can converge faster.
* Weight updates can remain stable.
* The model can reach a low-loss region efficiently.

---

## Limitations

### Learning Rate Too Small

```text
Training → Very Slow
```

### Learning Rate Too Large

```text
Training → Unstable
```

### Learning Rate Changes by Problem

There is no single learning rate that is perfect for every dataset and model.

---

## When to Use

Learning rate is important **whenever you train a model using gradient-based optimization**.

You should usually experiment with values such as:

```text
0.001
0.01
0.1
```

and observe the training behavior.

The suitable range depends on the optimizer and model.

---

## Important Point

Learning rate is **not the same as accuracy**.

For example:

```text
Learning Rate = 0.01
```

does **not** mean:

```text
Accuracy = 1%
```

It only controls the **size of the weight update**.

---

## Easy Memory Trick

Think about walking toward the bottom of a hill:

```text
Small Learning Rate
→ Small steps
→ Slow

Good Learning Rate
→ Proper steps
→ Efficient

Large Learning Rate
→ Huge steps
→ May overshoot
```

### One-line memory trick:

> **Learning Rate = How big a step should the model take when changing its weights?**


In [6]:

## Python Example

weight = 5.0
gradient = 2.0

learning_rates = [0.001, 0.01, 0.1]

for learning_rate in learning_rates:

    new_weight = weight - learning_rate * gradient

    print("Learning Rate:", learning_rate)
    print("New Weight:", new_weight)
    print()


Learning Rate: 0.001
New Weight: 4.998

Learning Rate: 0.01
New Weight: 4.98

Learning Rate: 0.1
New Weight: 4.8



```

## Output

```text
Learning Rate: 0.001
New Weight: 4.998

Learning Rate: 0.01
New Weight: 4.98

Learning Rate: 0.1
New Weight: 4.8
```

Look at the difference:

```text
0.001 → 4.998
0.01  → 4.98
0.1   → 4.8
```

The larger the learning rate, the larger the weight update.

---




In [8]:
## Another Python Example

#Let's actually train a small model using different learning rates.


import torch
import torch.nn as nn

X = torch.tensor([[1.0], [2.0], [3.0], [4.0]])
y = torch.tensor([[2.0], [4.0], [6.0], [8.0]])

learning_rates = [0.001, 0.01, 0.1]

for lr in learning_rates:

    model = nn.Linear(1, 1)

    optimizer = torch.optim.SGD(
        model.parameters(),
        lr=lr
    )

    loss_function = nn.MSELoss()

    for epoch in range(100):

        optimizer.zero_grad()

        predictions = model(X)

        loss = loss_function(predictions, y)

        loss.backward()

        optimizer.step()

    print(
        "Learning Rate:",
        lr,
        "Final Loss:",
        round(loss.item(), 6)
    )

Learning Rate: 0.001 Final Loss: 1.460299
Learning Rate: 0.01 Final Loss: 0.049574
Learning Rate: 0.1 Final Loss: 0.0


```

## What This Example Tells

We train the **same type of model** three times.

Only the learning rate changes:

```text
0.001
0.01
0.1
```

We can then observe how the final loss changes.

The exact results can vary because the model starts with randomly initialized weights.

---

# 7. SGD with Momentum

## Simple Definition

**SGD with Momentum** is an improved version of SGD that uses the **previous weight update** to help decide the next update.

👉 Simple idea:

**SGD:** Move based only on the current gradient.
**Momentum:** Move based on the current gradient + previous movement.

---

## Easy Explanation

Imagine you are pushing a heavy ball down a hill.

* Normal SGD → the ball reacts mainly to the current slope.
* SGD with Momentum → the ball has some **speed/inertia** from its previous movement.

So, instead of changing direction too quickly, momentum helps the model move more smoothly toward the minimum loss.

### Basic Flow

```text
Current Gradient
       ↓
Previous Movement
       ↓
Combine Both
       ↓
Update Weight
```

---

## Mathematical Intuition

Normal SGD:

```text
New Weight = Old Weight - Learning Rate × Gradient
```

With Momentum, we first calculate a **velocity**:

```text
Velocity = Momentum × Previous Velocity + Gradient
```

Then:

```text
New Weight = Old Weight - Learning Rate × Velocity
```

Usually, the momentum value is something like:

```text
0.9
```

This means the previous movement has a strong influence on the next movement.

---

## How It Updates Weights

Suppose:

```text
Old Weight = 5
Current Gradient = 2
Learning Rate = 0.1
Previous Velocity = 3
Momentum = 0.9
```

First:

```text
Velocity = (0.9 × 3) + 2
         = 4.7
```

Then:

```text
New Weight = 5 - (0.1 × 4.7)
           = 4.53
```

So the weight moves from:

```text
5 → 4.53
```

The previous movement helped make the current update larger.

---


---

## Real-World Example

### House Price Prediction

Suppose a neural network predicts house prices using:

* House size
* Number of bedrooms
* House age
* Location
* Distance from city center

During training, SGD changes these weights based on the current gradient.

Sometimes the gradient changes slightly from one batch to another.

Normal SGD may behave like:

```text
→ ↓ → ↗ ↓ → ↖ ↓
```

The movement can be noisy.

Momentum tries to make the movement more consistent:

```text
→ → → ↓ ↓ ↓
```

So the model can move more smoothly toward a lower-loss region.

---

## Advantages

* Can make training **faster**
* Reduces some of the noisy movement of normal SGD
* Helps move through long or shallow directions
* Can help the model pass through small local irregularities more effectively

---

## Limitations

* Has an additional parameter: `momentum`
* Learning rate still needs to be chosen carefully
* Very high momentum can cause overshooting
* More complicated than basic SGD

---

## When to Use

SGD with Momentum can be useful when:

* Training neural networks
* Basic SGD is moving too slowly
* Training is noisy
* You want the benefits of SGD with smoother/faster movement

Common starting values are:

```python
lr = 0.01
momentum = 0.9
```

These are starting points, not guaranteed best values.

---

## Key Points

```text
SGD
↓
Uses current gradient

SGD + Momentum
↓
Uses current gradient + previous movement
```

Remember:

* **Gradient** → tells which direction to move
* **Learning Rate** → tells how big the step is
* **Momentum** → remembers previous movement
* `optimizer.step()` → performs the weight update

### Easy Memory Trick

> **Momentum = Previous movement + Current gradient**


In [9]:

## Python Example
import torch

weight = torch.tensor(5.0, requires_grad=True)

optimizer = torch.optim.SGD(
    [weight],
    lr=0.1,
    momentum=0.9
)

for step in range(5):

    optimizer.zero_grad()

    loss = weight ** 2

    loss.backward()

    optimizer.step()

    print(
        f"Step {step + 1}: "
        f"Weight = {weight.item():.4f}, "
        f"Loss = {loss.item():.4f}"
    )


Step 1: Weight = 4.0000, Loss = 25.0000
Step 2: Weight = 2.3000, Loss = 16.0000
Step 3: Weight = 0.3100, Loss = 5.2900
Step 4: Weight = -1.5430, Loss = 0.0961
Step 5: Weight = -2.9021, Loss = 2.3808


```

### Output

You will see the weight moving toward `0` and the loss decreasing.

Example:

```text
Step 1: Weight = 4.0000, Loss = 25.0000
Step 2: Weight = 2.3000, Loss = 16.0000
Step 3: Weight = 0.3100, Loss = 5.2900
Step 4: Weight = -1.3780, Loss = 0.0961
Step 5: Weight = -2.4860, Loss = 1.8989
```

Here you can notice something important:

The weight **does not move smoothly toward zero every time**.

Because momentum carries previous movement, the model can **overshoot** the minimum and move to the other side.

With a suitable learning rate and momentum, this behavior can help the model reach a good solution faster.

# 8. RMSprop

## Simple Definition

**RMSprop** is an optimizer that automatically adjusts the **size of weight updates** based on the recent gradients.

👉 Simple idea:

**Large gradient → smaller step**
**Small gradient → larger step**

This helps the model avoid making very large updates in some directions.

---

## Easy Explanation

Imagine walking down a hill.

Suppose one direction has a **very steep slope**.

If you take the same large step there, you may jump past the lowest point.

RMSprop says:

> "This direction has been producing large gradients, so I should take smaller steps here."

For directions where gradients are small, RMSprop can take relatively larger steps.

So:

```text
Large gradient
      ↓
Reduce update size

Small gradient
      ↓
Allow larger update
```

---

## Mathematical Intuition

RMSprop keeps a **running average of squared gradients**.

Simplified:

```text
Average = β × Previous Average
        + (1 - β) × Gradient²
```

Then the weight update is approximately:

```text
New Weight =
Old Weight
- Learning Rate × Gradient / √(Average + ε)
```

Where:

* `β` → controls how much previous gradients are remembered
* `ε` → very small value used for numerical stability
* `Learning Rate` → controls the overall update size

You don't need to memorize the formula now.

The main idea is:

> **RMSprop changes the effective learning rate for each parameter.**

---

## How It Updates Weights

Suppose a model has two weights:

```text
Weight 1 → gradient = 10
Weight 2 → gradient = 1
```

A normal SGD optimizer applies the learning rate directly to both gradients.

RMSprop looks at the history of gradients.

Because Weight 1 has a much larger gradient, RMSprop reduces its effective step.

So conceptually:

```text
Weight 1
Large gradient
     ↓
Smaller effective step

Weight 2
Small gradient
     ↓
Larger effective step
```

This is called **adaptive learning rate**.

---

---

## Real-World Example

### Online Shopping Recommendation

Imagine an e-commerce model predicting whether a customer will buy a product.

The model may use:

* Previous purchases
* Product views
* Search history
* Product price
* Discount
* Customer activity

Some parameters may produce very large gradients during training, while others produce small gradients.

RMSprop adjusts the update for each parameter.

```text
Large gradient
      ↓
Smaller update

Small gradient
      ↓
Relatively larger update
```

This can make training more stable.

---

## Advantages

* Automatically adapts the learning rate
* Handles different gradient sizes well
* Can work well for neural networks
* Often useful when gradients vary significantly

---

## Limitations

* Has additional hyperparameters
* Can sometimes converge to a less desirable solution depending on the problem
* Adam is often preferred as a general starting point because it combines adaptive learning with momentum-like behavior

---

## When to Use

RMSprop can be useful when:

* Gradients have very different sizes
* Training is unstable with basic SGD
* Working with neural networks
* You want an adaptive learning-rate optimizer

It is also historically popular for some recurrent neural network and sequence-learning problems.

---

## Key Points

Remember this:

```text
SGD
↓
Same learning-rate idea for parameters

RMSprop
↓
Adjusts effective learning rate
based on recent gradient history
```

### Easy Memory Trick

> **RMSprop = Large gradient → smaller step**

And:

```text
Gradient
   ↓
RMSprop checks gradient history
   ↓
Adjusts update size
   ↓
Weight update
```


In [10]:

## Python Example

#PyTorch provides RMSprop directly:

import torch

weight = torch.tensor(5.0, requires_grad=True)

optimizer = torch.optim.RMSprop(
    [weight],
    lr=0.1
)

for step in range(5):

    optimizer.zero_grad()

    loss = weight ** 2

    loss.backward()

    optimizer.step()

    print(
        f"Step {step + 1}: "
        f"Weight = {weight.item():.4f}, "
        f"Loss = {loss.item():.4f}"
    )



Step 1: Weight = 4.0000, Loss = 25.0000
Step 2: Weight = 3.3734, Loss = 16.0000
Step 3: Weight = 2.9043, Loss = 11.3798
Step 4: Weight = 2.5283, Loss = 8.4351
Step 5: Weight = 2.2157, Loss = 6.3921


```

### Output

The exact values depend on the optimizer settings, but you should see the weight moving toward the minimum and the loss generally decreasing.

Example pattern:

```text id="rms02"
Step 1 → Weight decreases
Step 2 → Weight decreases
Step 3 → Weight decreases
Step 4 → Weight decreases
Step 5 → Weight decreases
```

The important thing is not memorizing the exact numbers.

The important thing is understanding that **RMSprop adjusts the update size using gradient history**.

# 9. Adam Optimizer

## Simple Definition

**Adam (Adaptive Moment Estimation)** is an optimizer that combines ideas from **Momentum** and **RMSprop**.

👉 Simple idea:

* **Momentum** → remembers previous gradient movement.
* **RMSprop** → adjusts the step size based on gradient size.
* **Adam** → uses both ideas together.

That is why Adam is very commonly used for training neural networks.

---

## Easy Explanation

Think about training a neural network like driving toward a destination.

Adam looks at two things:

### 1. Which direction have I been moving?

This comes from **Momentum**.

### 2. How large are the gradients?

This comes from the **RMSprop idea**.

Then Adam decides:

> "Based on my previous direction and the size of the gradients, how much should I change this weight?"

So:

```text
Current Gradient
       +
Previous Gradient Information
       +
Gradient Size Information
       ↓
      Adam
       ↓
Weight Update
```

---

## Mathematical Intuition

Adam maintains two important values:

### First Moment

This tracks the **average direction of gradients**.

```text
m = moving average of gradients
```

This is similar to **Momentum**.

### Second Moment

This tracks the **average size of squared gradients**.

```text
v = moving average of squared gradients
```

This is similar to **RMSprop**.

The simplified Adam update is:

```text
New Weight =
Old Weight
- Learning Rate × m / √(v + ε)
```

You don't need to memorize the complete Adam equations right now.

Just remember:

```text
Momentum → remembers direction

RMSprop → adjusts step size

Adam → combines both
```

---

## How It Updates Weights

Suppose a neural network has a weight:

```text
Weight = 5
```

During training, Adam receives gradients from the current batch.

Adam considers:

```text
1. Current gradient
2. Previous gradient information
3. Size of recent gradients
4. Learning rate
```

Then it calculates an appropriate update.

Conceptually:

```text
Weight = 5.0
   ↓
Adam calculates update
   ↓
Weight = 4.7
   ↓
Next training step
   ↓
Weight = 4.3
   ↓
Next training step
   ↓
Weight = 3.8
```

The exact values depend on the model, gradients, learning rate, and Adam's internal state.

---


## Real-World Example

### Hospital Patient Risk Prediction

Suppose a neural network predicts whether a patient is at high risk using:

* Age
* Blood pressure
* Blood glucose
* Heart rate
* Previous medical measurements
* Laboratory results

During training, different parameters can have very different gradient sizes.

Adam can:

* Remember the direction in which the model has been moving.
* Adjust the update size for different parameters.
* Update the network weights automatically.

So training becomes:

```text
Patient Data
     ↓
Neural Network
     ↓
Prediction
     ↓
Loss
     ↓
Gradients
     ↓
Adam
     ↓
Updated Weights
```

---

## Advantages

* Combines momentum and adaptive learning-rate ideas
* Usually requires less manual tuning than basic SGD
* Works well for many neural-network problems
* Handles parameters with different gradient sizes
* Commonly used as a starting optimizer

---

## Limitations

* Uses additional memory to store optimizer information
* Still depends on a suitable learning rate
* It is not automatically the best optimizer for every problem
* For some models, SGD with Momentum may have desirable training behavior

---

## When to Use

Adam is a good option when:

* You are training neural networks
* You want an adaptive optimizer
* You are starting a new deep-learning project
* Basic SGD training is difficult to tune

A common starting point is:

```python id="adam03"
optimizer = torch.optim.Adam(
    model.parameters(),
    lr=0.001
)
```

`0.001` is a common starting value, **not a rule**.

---

## Key Points

### Remember the three optimizers

```text
SGD
↓
Uses gradient + learning rate


SGD + Momentum
↓
Gradient + previous movement


RMSprop
↓
Gradient + adaptive step size


Adam
↓
Momentum idea + RMSprop idea
```

### Easy Memory Trick

> **Adam = Momentum + Adaptive Learning**

And the basic training flow is:

```text
Forward Pass
      ↓
Loss
      ↓
loss.backward()
      ↓
Gradients
      ↓
Adam
      ↓
optimizer.step()
      ↓
Updated Weights
```


In [12]:

## Python Example

#PyTorch provides Adam directly:

import torch

weight = torch.tensor(5.0, requires_grad=True)

optimizer = torch.optim.Adam(
    [weight],
    lr=0.1
)

for step in range(5):

    optimizer.zero_grad()

    loss = weight ** 2

    loss.backward()

    optimizer.step()

    print(
        f"Step {step + 1}: "
        f"Weight = {weight.item():.4f}, "
        f"Loss = {loss.item():.4f}"
    )


Step 1: Weight = 4.9000, Loss = 25.0000
Step 2: Weight = 4.8001, Loss = 24.0100
Step 3: Weight = 4.7002, Loss = 23.0406
Step 4: Weight = 4.6005, Loss = 22.0920
Step 5: Weight = 4.5010, Loss = 21.1647


# 10. SGD vs Adam

## Simple Definition

**SGD** and **Adam** are both optimizers used to update neural-network weights.

The main difference is:

```text
SGD
↓
Uses the gradient directly
```

```text
Adam
↓
Uses gradient history
+ adaptive update size
```

---

## Easy Explanation

Imagine you are walking down a hill to reach the lowest point.

### SGD

SGD looks at the current slope and takes a step.

```text
Current slope
     ↓
How should I move?
     ↓
Take a step
```

### Adam

Adam also remembers previous movement and considers how large the gradients have been.

```text
Current slope
     +
Previous movement
     +
Gradient size history
     ↓
Decide the step
```

So Adam has more information when deciding how to update weights.

---

## Main Difference

| Feature                                 | SGD                   | Adam                       |
| --------------------------------------- | --------------------- | -------------------------- |
| Uses gradient                           | Yes                   | Yes                        |
| Remembers previous gradient information | Momentum version does | Yes                        |
| Adaptive step size                      | No                    | Yes                        |
| Number of internal values               | Lower                 | Higher                     |
| Memory usage                            | Lower                 | Higher                     |
| Learning-rate sensitivity               | Usually higher        | Often easier to start with |
| Common use                              | Neural networks       | Neural networks            |

---

## Mathematical Intuition

### SGD

```text
New Weight =
Old Weight
- Learning Rate × Gradient
```

Example:

```text
Old Weight = 5
Gradient = 2
Learning Rate = 0.1
```

Therefore:

```text
New Weight = 5 - (0.1 × 2)
           = 4.8
```

---

### Adam

Adam considers gradient history and gradient magnitude.

Conceptually:

```text
Gradient
   +
Previous gradient information
   +
Gradient magnitude information
   ↓
Adam
   ↓
Weight update
```

So Adam does **more processing** before deciding the update.

---

## Python Example

Let's create the same simple model and train it using both optimizers.

```python id="compare01"
import torch
import torch.nn as nn

# Training data
X = torch.tensor([[1.0], [2.0], [3.0], [4.0]])
y = torch.tensor([[2.0], [4.0], [6.0], [8.0]])

# SGD model
model_sgd = nn.Linear(1, 1)

optimizer_sgd = torch.optim.SGD(
    model_sgd.parameters(),
    lr=0.01
)

# Adam model
model_adam = nn.Linear(1, 1)

optimizer_adam = torch.optim.Adam(
    model_adam.parameters(),
    lr=0.01
)

loss_function = nn.MSELoss()

for epoch in range(100):

    # ----- SGD -----
    optimizer_sgd.zero_grad()

    prediction_sgd = model_sgd(X)

    loss_sgd = loss_function(prediction_sgd, y)

    loss_sgd.backward()

    optimizer_sgd.step()


    # ----- Adam -----
    optimizer_adam.zero_grad()

    prediction_adam = model_adam(X)

    loss_adam = loss_function(prediction_adam, y)

    loss_adam.backward()

    optimizer_adam.step()


print("Final SGD Loss:", loss_sgd.item())
print("Final Adam Loss:", loss_adam.item())
```

### Output

Your exact values may be slightly different because the models start with different random weights.

You should generally see both losses decrease:

```text
Final SGD Loss: small value
Final Adam Loss: small value
```

The important observation is:

> Both SGD and Adam learn by using gradients, but they calculate their weight updates differently.

---

## Real-World Example

### E-Commerce Purchase Prediction

Suppose we build a neural network to predict whether a customer will purchase a product.

Input data:

* Previous purchases
* Product views
* Search activity
* Discount viewed
* Time spent on product page

The model produces:

```text
Prediction
    ↓
Loss
    ↓
Gradient
    ↓
Optimizer
    ↓
Updated weights
```

If we use SGD:

```text
Gradient
   ↓
SGD
   ↓
Weight update
```

If we use Adam:

```text
Gradient
   ↓
Gradient history
   +
Adaptive step calculation
   ↓
Adam
   ↓
Weight update
```

Both can be used to train the same neural network.

---

## Advantages of SGD

* Simple
* Uses less optimizer memory
* Gives direct control over the learning process
* SGD with Momentum can improve training movement

## Limitations of SGD

* Learning rate can require careful tuning
* Training can be noisy
* May require momentum or learning-rate scheduling

---

## Advantages of Adam

* Adaptive updates
* Uses gradient history
* Often convenient as a starting optimizer
* Can work well across many neural-network problems

## Limitations of Adam

* Uses more memory than basic SGD
* Still requires hyperparameter tuning
* More complicated internally

---

## When to Use

### SGD

Consider SGD when:

```text
You want:
Simple optimizer
+
Lower optimizer memory
+
More direct control
```

### Adam

Consider Adam when:

```text
You want:
Adaptive updates
+
Gradient history
+
A convenient starting point
```

There is no universal rule that one optimizer is always better. The choice depends on the model, dataset, learning-rate setup, and training behavior.

---

## Key Points

Remember this simple comparison:

```text
SGD
↓
Current gradient
↓
Update weight
```

```text
Adam
↓
Current gradient
+
Gradient history
+
Adaptive step size
↓
Update weight
```

### Easy Memory Trick

> **SGD = Simple update**

> **Adam = Adaptive update**

And in PyTorch:

```python
torch.optim.SGD(...)
```

and

```python
torch.optim.Adam(...)
```

are both used as the optimizer in:

```text
Forward Pass
      ↓
Loss
      ↓
Backward Pass
      ↓
Gradients
      ↓
Optimizer
      ↓
```
